persistence/rest/src/main/java/org/infinispan/persistence/rest/RestStore.java

```diff
@@ -365,37 +276,14 @@ private long timeHeaderToLong(String header) {
    @Override
    public Flowable<K> publishKeys(Predicate<? super K> filter) {
       return Flowable.using(() -> {
-         DefaultHttpRequest get = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, path + "?global");
-         get.headers().add(HttpHeaderNames.ACCEPT, "text/plain");
-         get.headers().add(HttpHeaderNames.ACCEPT_CHARSET, "UTF-8");
-         HttpResponseHandler handler = new HttpResponseHandler(true);
-         try {
-            Channel ch = bootstrap.connect(configuration.host(), configuration.port()).awaitUninterruptibly().channel().pipeline().addLast(
-                  new HttpObjectAggregator(maxContentLength), handler).channel();
-            ch.writeAndFlush(get).sync().channel().closeFuture().sync();
-            BufferedReader reader = new BufferedReader(new InputStreamReader(new ByteBufInputStream((handler.getResponse()).content()), StandardCharsets.UTF_8));
-            // Response can't be null now
-            return new KeyValuePair<>(handler.getResponse(), reader);
-         } catch (Throwable t) {
-            FullHttpResponse response = handler.getResponse();
-            if (response != null) {
-               response.release();
-            }
-            throw t;
-         }
-      }, kvp -> Flowable.fromIterable(() -> new RestIterator(kvp, filter)), kvp -> {
-         try {
-            kvp.getValue().close();
-         } finally {
-            kvp.getKey().release();
-         }
-      });
+         RestResponse response = CompletableFutures.await(cacheClient.keys().toCompletableFuture());
+         return new BufferedReader(new InputStreamReader(response.getBodyAsStream()));
+      }, kvp -> Flowable.fromIterable(() -> new RestIterator(kvp, filter)), BufferedReader::close);
```

Comment:
IMO it's ok to wait for the response in the blocking methods like `write()`, but `publishX` methods should not block, especially not when the caller subscribes.

Perhaps something like this:

```
      return Flowable.using(() -> cacheClient.keys(),
                            stage -> RxJavaInterop.completionStageToSingle(stage)
                                                  .flatMapPublisher(response -> Flowable.fromIterable(...),
                            stage -> { ??? })
```

I've put question marks in the disposal section because I'm not sure what's the best approach. After we have the input stream it should be enough to close that, but before the stage is done we don't have any cancel/close method. Perhaps `stage -> stage.thenAccept(response -> response.getBodyAsStream().close())`?

/cc @wburns

Reply:
This behaviour was already part of the code, but I can look into this...

Reply:
The ```entryPublisher``` method  also blocks

Reply:
You marked the conversation as resolved and I missed your answers :)

But you're right, this PR is already doing enough, so I don't mind handling the blocking inside reactive streams later.